haserl.1
.\" Process with groff -man -Tascii haserl.1
.TH haserl 1 "August 2024"
.SH NAME
haserl \- A CGI scripting program for embedded environments
.SH SYNOPSIS
.BI "#!/usr/bin/haserl [\-\-upload\-dir=" dirspec "] [\-\-upload\-limit=" limit "]"

.SH DESCRIPTION
Haserl is a small CGI wrapper that uses Lua as the programming language. It is
very small, so it can be used in embedded environments, or where something like
PHP is too big.

.SH OPTIONS

.TP
\fB\-U\fR, \fB\-\-upload\-dir=\fIdirspec\fR
Defaults to "/tmp". All uploaded files are created with temporary filename in
this directory
.B FORM.xxx_path
contains the name of the temporary file.
.B FORM.xxx_filename
contains the original name of the file, as specified by the client.

.TP
\fB\-u\fR, \fB\-\-upload\-limit=\fIlimit\fR
Allow a mime-encoded file up to
.I limit KB
to be uploaded. The default is
.I 0KB
(no uploads allowed).
Note that mime-encoding adds 33% to the size of the data.

.SH OVERVIEW OF OPERATION

In general, the web server sets up several environment variables, and then uses
.I fork
or another method to run the CGI script. If the script uses
.IR haserl ,
the following happens:

.IP
If
.I haserl
is run as root, then the uid/gid is set to the owner/group of the script.

The provided cookie and GET/POST data is parsed. Form-elements are placed as
name=value pairs into the Lua environment for the CGI script to use.

The Lua script is executed.

.SH CLIENT SIDE INPUT
.I haserl
will decode data sent via the
.I HTTP_COOKIE
environment variable, and the GET or POST method from the client, and store them
as elements in tables that can be accessed by the executed Lua script. The name
of the variable corresponds to the name of the method.

For cookies, elements are stored in the
.I COOKIE
variable. For example, if
.I HTTP_COOKIE
includes "foo=bar", then:
.B COOKIE.foo
== "bar".

For the GET method, data sent in the form %xx is translated into the characters
they represent, and elements are stored in the
.I GET
variable. For example, if
.I QUERY_STRING
includes "foo=bar", then:
.B GET.foo
== "bar" .

For the POST method, elements are stored in the
.I POST
variable. For example, if the post stream includes "foo=bar", then:
.B POST.foo
== "bar" .

Also, for the POST method, if the data is sent using
.I "multipart/form\-data"
encoding, the data is automatically decoded. This is typically used when files
are uploaded from a web client using <input type=file>.

.TP
.B NOTE
When a file is uploaded to the web server, it is stored in the
.I upload-dir
directory.
.B FORM.variable_filename
contains the name of the file uploaded (as specified by the client).
.B FORM.variable_path
contains the name of the file in
.I upload-dir
that holds the uploaded content. To prevent malicious clients from filling up
.I upload-dir
on your web server, file uploads are only allowed when the
.I \-\-upload\-limit
option is used to specify how large a file can be uploaded. Haserl automatically
deletes the temporary file when the script is finished. To keep the file, move
it or rename it somewhere in the script.

.P
When multiple instances of the same variable are sent via the same method, only
the last one is saved.

.I haserl
redefines the default Lua
.I print
function in order to buffer its output. This effectively protects against
partially executed responses when an error occurs. If the script were to
interact directly with standard output, any output from the script would
interfere with
.IR haserl 's
own error messages and produce garbled output.
.br
For the sake of programmer
convenience, the arguments of
.I print
mirror those of
.IR string.format .
Consult the sections below for usage examples.

.SH EXAMPLES
.TP
.B WARNING
The examples below are simplified to show how to use
.IR haserl .
You should be familiar with basic web scripting security before using
.I haserl
(or any scripting language) in a production environment.

.TP
.B Looping with dynamic output
.nf
#!/usr/bin/haserl
print("Content-type: text/html\\r\\n\\r\\n")

print("<html><body><table border=1><tr>")
for i, v in ipairs({"Red", "Blue", "Yellow", "Cyan"}) do
	print("<td bgcolor='%s'>%s</td>", v, v)
end
print("</tr></table></body></html>")
.fi

Sends a mime-type "text/html" document to the client, with a table
containing elements labeled with each respective background color.

.TP
.B Self Referencing CGI with a form
.nf
#!/usr/bin/haserl
print("Content-Type: text/html\\r\\n\\r\\n")

print("<html><body>")
print("<h1>Sample Form</h1>")
print("<form action='%s' method='GET'>", os.getenv("SCRIPT_NAME"))

-- Do some basic validation of POST.textfield
-- To prevent common web attacks
print("<input type=text name=textfield Value='%s' cols=20>",
      POST.textfield:gsub("[^%w ]", ""):gsub("%l+", string.upper))
print("<input type=submit value=GO>")

print("</form>")
print("</body></html>")
.fi

Prints a form. If the client enters text in the form, the CGI is reloaded
(defined by the
.IR SCRIPT_NAME
environment variable) and the textfield is sanitized to prevent web attacks,
then the form is redisplayed with the text the user entered. The text is
uppercased.

.TP
.B Uploading a File
.nf
#!/usr/bin/haserl \-\-upload\-limit=4096
print("Content-Type: text/html\\r\\n\\r\\n")

print("<html><body>")
print("<form action='%s' method=POST enctype='multipart/form-data'>", os.getenv("SCRIPT_NAME"))
print("<input type=file name=uploadfile>")
print("<input type=submit value=GO>")
print("<br>")
if FORM.uploadfile_path then
	print("<p>You uploaded a file named <b>%s</b>, and it was", FORM.uploadfile_filename)
	print("temporarily stored on the server as <i>%s</i>.", FORM.uploadfile_path)
	local file = io.open(FORM.uploadfile_path, "r")
	if file then
		local size = file:seek("end")
		file:close()
		print("The file was %d bytes long.</p>", size)
		os.remove(FORM.uploadfile_path)
		print("<p>Don't worry, the file has just been deleted from the web server.</p>")
	end
else
	print("You haven't uploaded a file yet.")
end
print("</form>")
print("</body></html>")
.fi

Displays a form that allows for file uploading. This is accomplished by using the
.B \-\-upload\-limit
and by setting the form
.I enctype
.RI "to " multipart/form\-data.
If the client sends a file, then some information regarding the file is printed,
and then deleted. Otherwise, the form states that the client has not uploaded a
file.

.SH SAFETY FEATURES
.TP
.B Privilege Dropping
If run as root (usually due to being installed with suid),
.I haserl
will set its uid/gid to that of the owner/group of the script. This can be used
to have a set of CGI scripts where each has a different privilege level. If the
.I haserl
binary is not run as root, then the CGI scripts will run with the uid/gid of the
web server.

.TP
.B Reject command line parameters given on the URL
If the URL does not contain an unencoded "=", then the CGI spec states the
options are to be used as command-line parameters to the program. For instance,
according to the CGI spec:
.I http://192.168.0.1/test.cgi?\-\-upload\-limit%3d2000&foo%3dbar
.br
Should set the upload-limit to 2000KB in addition to setting "Foo=bar". To
protect against clients enabling their own uploads,
.I haserl
rejects any command-line options beyond argv[2]. If invoked as a #! script, the
interpreter is argv[0], all command-line options listed in the #! line are
combined into argv[1], and the script name is argv[2].

.SH LUAC

.I luac
can be used to produce a precompiled Lua chunk. All Lua features listed above
are still available.

Here is an example of a trivial script, converted into a luac CGI script:

Given the file test.lua:
.RS
.nf
print("Content-Type: text/plain\\r\\n\\r\\n")
print("Hello, world!\\r\\n")
.fi
.RE

It can be compiled with luac:
.RS
luac \-o test.luac \-s test.lua
.RE

And then a wrapper script to call it:
.RS
.nf
#!/bin/sh
exec haserl test.luac
.fi
.RE

.SH NAME
The name "haserl" comes from the Bavarian word for "bunny." At first glance it
may be small and cute, but
.I haserl
is more like the bunny from
.IR "Monty Python & The Holy Grail" .
In the words of Tim the Wizard,
.I That's the most foul, cruel & bad-tempered rodent you ever set eyes on!

Haserl can be thought of the CGI equivalent to
.IR netcat .
Both are small, powerful, and have very little in the way of extra features. Like
.IR netcat ", " haserl
attempts to do its job with the least amount of extra "fluff".

.SH AUTHORS
neeshy <neeshy@tfwno.gf>
.br
Nathan Angelacos <nangel@users.sourceforge.net>

.SH SEE ALSO
haserl (the original) (https://haserl.sourceforge.net/)
.br
uncgi (http://www.midwinter.com/~koreth/uncgi.html)
.br
cgiwrapper (http://cgiwrapper.sourceforge.net/)


ChangeLog
2015-03-11
	0.9.35
	*	Geert Jan de Groot reported a possible segfault if CONTENT_TYPE is not
		specified.  This bug was introduced in 0.9.33

2015-02-15
	0.9.34
	*	Fixes for Lua 5.3 compatibility
	*	Move sys/fnctl.h to fnctl.h
	*	Victor Cook sent in a patch to rfc2388.c -
		"In rfc2388.c rfc2388_handler() during processing of the header information,
		if a buffer read falls on the point where the next character is crlf it
		results in the header reading to be terminated early. When the error occurs
		the results are unpredictable and depend on what is then incorrectly interpreted.

		The chance of a failure depends on the length and complexity of the header data,
		but can become very significant."

2014-06-07
	0.9.33
	*	Fix various security vulnerabilities - most serious is a
		  Heap Overflow Vulnerability in sliding_buffer.c
		Reported by Stephen Röttger
	*	Allow PUT and DELETE method (But prefix is still POST/GET)
	*	On POST/PUT, Content-Type is not x-www-urlencoded, then
		the body of the message is stored verbatim in POST_body=


2013-20-09
	0.9.32
	*	Regression causing Lua always to be linked, never used

2013-15-09
	0.9.31
	*	Modernize configure.ac
	*	more Lua 5.2 fixes


2013-27-06
	0.9.30
	*	The Mayhem Team of CMU found an undisclosed segfault when the first
		command-line argument is '' or "" (null-quoted string).  Chow Loong Jin
		supplied a patch.
	*	Natanael Copa supplied a patch to rename the deprecated string.gfind
		in haserl_lualib.lua.  Lua 5.2 is now supported.

2011-05-09
	0.9.29
	*	Jan Rome reported an off-by one error in handling array input (formvar[])
	*	Chow Loong Jin reported multiple header files with no copyright information
		This is part of the process to get haserl into Ubuntu.
	*	Chow Loong Jin noticed haserl_lua.inc is no longer necessary in the
		tarball  (sf id 3298883)
	*	Oliver Metz noted that haserl aborted on SIGPIPE from a custom upload
		handler.  Haserl now ignores the signal.
		(Example: uploading to a 'tar xvf' handler)
		text       data     bss     dec     hex filename
		19637       836     176   20649    50a9 src/haserl

2011-03-25
	0.9.28
	*	Lua script did not have variables in the correct tables (Mixure for
		FORM. and POST_, GET_ They should now be in the correct table format
	*	Natanael Copa provided patches to compile lua2c if necessary
	*	Chow Loong Jin proposes using printf '%s' instead of echo -n;
		no reason not to.
		text       data     bss     dec     hex filename
		28598       948     176   29722    741a src/haserl

2010-10-17
	0.9.27
	*	Daniel Griscom reported clarification on RFC 2616 compliance
		Ralph Siemsen provided the actual man-page update
	*	Natanael Copa fixed the Makefile.am for gnu make 3.82+
	*	Cleanup temp files on error (Anonymous) SF bug tracker id 2991410
	*	Pavel Chromy reported a vulnerability where the name of a the tempfile
		on a file upload could be modified by the client
	*	man page updates
		text       data     bss     dec     hex filename
		19343       828     172   20343    4f77 src/haserl

2009-07-30
	0.9.26
	*	Peter Korsgaard found a buffer overflow error in handling
		array variables from the client.
	*	Changed buffer_init function to haserl_buffer_init to
		work around a nasty problem when using luasocket.
		http://lua-users.org/lists/lua-l/2008-08/msg00446.html
		text       data     bss     dec     hex filename
		18873       820     172   19865    4d99 src/haserl
2008-12-16
	0.9.25
	*	Somehow version control failed, and the -d debug short
		option reverted to -D again.  fixed.
	*	haserl.c - "command-line" handling was broken on OSX and BSD
		fixed. (Mark Blythe)
	*	haserl.c - fix bug where CONTENT_LENGTH=0 would hang haserl
		(bug tracker bug #1959379)
		text       data     bss     dec     hex filename
		18949       820     172   19941    4de5 haserl
2008-04-14
	0.9.24
	*	haserl.c - myputenv caused a segfault when a variable without
		an = (assignment) was passed.  (Scott)
	*	doc/haserl.1 - Misc typos fixed (Scott)
	*	scott's bash-extensions are included if configured with
		--enable-bash-extensions
	*	running haserl with no args now tells you if lua or
		the bash extensions are enabled.
	*	configure.ac - FORM_ now hardcoded
	*	tests/* - beginnings of unit tests (make check)
	*	haserl.c - varibles now stored as FORM_ + GET_, POST_, COOKIE_
		(scott) - WIP - need to clean up the putenv so that only one
		copy is saved in the host, and the child gets all the
		duplicates
		text       data     bss     dec     hex filename
		18519       824     172   19515    4c3b src/haserl

2008-03-22
	0.9.23
	*	haserl.c - short option for debug id -d (was incorrectly -D)
	*	haserl.c - remove use of legacy "index" clib function.
	*	haserl.c - myputenv makes a newline delimited "array" out
		of variables that end in [].  (php-style for multi-selects)
		Lua variables will have same behavior - they are not treated
		as true arrays yet. (WIP - need to handle the env variables
		as abstracts until subshell startup for this to work)
		text       data     bss     dec     hex filename
		18237       820     128   19185    4af1 src/haserl
2008-01-28
	0.9.22
	*	h_bash.c - close the input side of the pipe, so that if the
		child dies early the parent doesn't deadlock (patch by Diego
		Santa Cruz)
	*	rfc2388.c, sliding_buffer.c, haserl.c - read CONTENT_LENGTH,
		and stop reading input when CONTENT_LENGTH reached (fixes odd
		problems where the client may not close the connection
		properly at the end of a request [guess which clients are
		MiStaken?])  (patch by Diego Santa Cruz)
	*	Allow "-" to be used in a field identifier (e.g. FORM["-foo"])
		text       data     bss     dec     hex filename
		17813       812     128   18753    4941 src/haserl
2007-11-22
	0.9.21
	*	haserl.c - fixed a stupid test for strlen that caused CGI post
		requests to take exponentially longer amounts of time based on
		the size of the upload
	*	h_bash.c - removed open_bash function; found that a CGI post

haserl.c
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>

#include <lua.h>

#include "common.h"
#include "buffer.h"

static int
unescape(char *where, const char *what)
{
#define decode() \
	if (c >= '0' && c <= '9') { \
		w |= c - '0'; \
	} else if (c >= 'A' && c <= 'F') { \
		w |= c - 'A' + 10; \
	} else if (c >= 'a' && c <= 'f') { \
		w |= c - 'a' + 10; \
	} else { \
		return 0; \
	}

	char c = what[0];
	char w = 0;
	decode()

	c = what[1];
	w <<= 4;
	decode()

	*where = w;

	return 1;
#undef decode
}

/* decode %xx to the bytes they represent */
static size_t
unescape_url(char *url)
{
	char *ptr = url;
	size_t i = 0;
	while (ptr[i]) {
		if ((*ptr = ptr[i]) == '%' && unescape(ptr, ptr + i + 1)) {
			i += 2;
		}
		ptr++;
	}
	*ptr = 0;

	return ptr - url;
}

/* add a pair of strings delimited on '=' to a buffer */
static void
lua_add_pair(const char *tbl, char *str)
{
	char *value = strchr(str, '=');
	if (value) {
		*value = 0;
		size_t key_size = unescape_url(str);
		size_t value_size = unescape_url(++value);
		lua_set(tbl, str, key_size, value, value_size);
	} else {
		size_t size = unescape_url(str);
		lua_set(tbl, str, size, "", 0);
	}
}

static void
read_cookie(const char *tbl, char *cookie)
{
	/* split on ; to extract name value pairs */
	char *token = strtok(cookie, ";");
	while (token) {
		/* skip leading spaces */
		while (*token == ' ') token++;
		lua_add_pair(tbl, token);
		token = strtok(NULL, ";");
	}
}

static void
read_query(const char *tbl, char *query)
{
	/* change pluses into spaces */
	for (char *s = query; *s; s++) {
		if (*s == '+') *s = ' ';
	}

	/* split on & to extract name value pairs */
	char *token = strtok(query, "&");
	while (token) {
		lua_add_pair(tbl, token);
		token = strtok(NULL, "&");
	}
}

/* read CGI variables from stdin (for POST queries) */
static void
read_form(void)
{
	/* if upload_limit is zero, we die right here */
	if (!global.upload_max) {
		die("File uploads are not allowed");
	}

	size_t max_len;
	char *content_length = getenv("CONTENT_LENGTH");
	if (!content_length) {
		die("Content length not specified");
	} else if (!(max_len = strtoul(content_length, NULL, 10))) {
		die("Invalid content length");
	} else if (max_len > global.upload_max) {
		die("Content length larger than allowed limits");
	}

	char *content_type = getenv("CONTENT_TYPE");
	if (content_type && !strncasecmp(content_type, "multipart/form-data", 19)) {
		multipart_handler();
		return;
	}

	buffer_t buf;
	buffer_alloc(&buf, CHUNK_SIZE);

	ssize_t n = read(0, buf.ptr, CHUNK_SIZE);
	while (n > 0) {
		buf.ptr += n;

		/* maximum size for non-multipart/form-data requests is CHUNK_SIZE */
		if (buf.ptr - buf.data >= CHUNK_SIZE) {
			buffer_destroy(&buf);
			die("Reached maximum allowed input length");
		}

		n = read(0, buf.ptr, buf.limit - buf.ptr);
	}

	if (n == -1) {
		buffer_destroy(&buf);
		die_status(errno, "read: %s", strerror(errno));
	}

	if (content_type && !strncasecmp(content_type, "application/x-www-form-urlencoded", 33)) {
		/* add the ASCIIZ */
		buffer_add(&buf, "", 1);
		read_query("POST", buf.data);
	} else {
		/* treat input as an opaque octet stream */
		lua_set("POST", "body", 4, buf.data, buf.ptr - buf.data);
	}
	buffer_destroy(&buf);

}

/* read the request data */
void
haserl(void)
{
	char *cookie = getenv("HTTP_COOKIE");
	if (cookie) {
		cookie = xstrdup(cookie);
		read_cookie("COOKIE", cookie);
		free(cookie);
	}

	char *request_method = getenv("REQUEST_METHOD");
	if (request_method) {
		if (!strcasecmp(request_method, "GET") ||
		    !strcasecmp(request_method, "DELETE")) {
			char *query = getenv("QUERY_STRING");
			if (query) {
				query = xstrdup(query);
				read_query("GET", query);
				free(query);
			}
		} else if (!strcasecmp(request_method, "POST") ||
		           !strcasecmp(request_method, "PUT")) {
			read_form();
		}
	}
}
